tests/functional/test_foreign_references.py

```diff
@@ -0,0 +1,220 @@
+
+
+@pytest.mark.asyncio
+async def test_fetch() -> None:
+    """Test that we can resolve the reference for a foreign key"""
+    org_id = str(uuid4())
+    serial_id = 104
+
+    try:
+        org = await Org.create(
+            id=org_id, serial_id=serial_id, name="fake org104", slug="fake slug104"
+        )
+        app = await App.create(
+            id=str(uuid4()),
+            name="app name",
+            org=org,
+            slug="app-slug",
+            data_cache_schema="app_ds1",
+            data_store_schema="app_dc1",
+        )
+
+        app_get = await App.get(app.id)
+        assert app_get.org.id == org.id
+        await app_get.org.fetch()
+        assert app_get.org == org
+    finally:
+        await org.delete()
+        await app.delete()
+
+
+@pytest.mark.asyncio
+async def test_swap_foreign_reference() -> None:
+    """Test that we can swap foreign key references"""
+    org_id = str(uuid4())
+    serial_id = 104
+
+    try:
+        org1 = await Org.create(
+            id=org_id, serial_id=serial_id, name="fake org104", slug="fake slug104"
+        )
+        org2 = await Org.create(
+            id=str(uuid4()),
+            serial_id=serial_id + 1,
+            name="fake org105",
+            slug="fake slug105",
+        )
+
+        # Start with app pointing to the first org
+        app = await App.create(
+            id=str(uuid4()),
+            name="app name",
+            org=org1,
+            slug="app-slug",
+            data_cache_schema="app_ds1",
+            data_store_schema="app_dc1",
+        )
+
+        # Confirm app is associated with org 1
+        app_get = await App.get(app.id)
+        assert app_get.org.id == org1.id
+
+        # Move app to org 2
+        app_get.org = org2
+        await app_get.save()
+        app_get = await App.get(app.id)
+        assert app_get.org.id == org2.id
+        await app_get.org.fetch()
+        assert app_get.org == org2
+    finally:
+        await org1.delete()
+        await org2.delete()
+        await app.delete()
+
+
+@pytest.mark.asyncio
+async def test_resolve_on_get() -> None:
+    """Test that we can resolve foreign keys when we retrieve the record object"""
+    org_id = str(uuid4())
+    serial_id = 104
+
+    try:
+        org = await Org.create(
+            id=org_id, serial_id=serial_id, name="fake org104", slug="fake slug104"
+        )
+        app = await App.create(
+            id=str(uuid4()),
+            name="app name",
+            org=org,
+            slug="app-slug",
+            data_cache_schema="app_ds1",
+            data_store_schema="app_dc1",
+        )
+
+        app_get = await App.get(app.id, resolve_references=True)
+        assert app_get.org == org
+    finally:
+        await org.delete()
+        await app.delete()
+
+
+@pytest.mark.asyncio
+async def test_nested_foreign_references() -> None:
+    """Test that we can nest foreign key references and resolve them"""
```

Comment:
If you get the topic with resolve_ref, then look at that app, do you still have to call fetch on the org to get it to pull down?

That is. if you added to this test, `topic_get = await Topic.get(topic.id, resolve_references=True)` and then asked for `topic_get.app.org.id` would you get the value?

Reply:
you would get that value but you wouldn't get `topic_get.app.org.slug` for instance. You can always get the `id` prop but the others won't get pulled.

We could change `get` with `resolve_references=True` to propagate that flag down however I was worried about performance if you have severely nested references. Open to changing it though